### DISCUSSION TITLE
Is it possible to find out when a download is done?

### DISCUSSION THREAD

**Discussion by KnorrFG** [tokio-rs/axum#2532] [Q&A] [answered]
### Summary

Hey, I would like to somehow know, when axum is done with a `StreamBody`, because I would like to put a Read-lock onto the source of the stream body that can be lifted again after a download.

I tried to create a wrapper Stream type that holds the lock handle and drops it when it itself is dropped, but I couldn't make it work so far. Is there an easier approach?

### axum version

0.6.20

**Comment by davidpdrsn** [comment, marked as answer]
> I tried to create a wrapper Stream type that holds the lock handle and drops it when it itself is dropped, but I couldn't make it work so far. Is there an easier approach?

That is the approach I recommend. Have a look at https://github.com/tokio-rs/axum/discussions/1060 which is basically the same question.

**Reply by KnorrFG** [reply]
This worked, thank you. Didn't know `async-stream`. Great crate.
